[Frederick Tang] iP

fredtwt opened the pull request:
# Doge project template

Its name is **_Doge_**, a person who occasionally ~~pisses~~ annoys you but at the end of the day, you still love them. You can download me [here](https://drive.google.com/file/d/1awN6tT8uNJPXamfNgMrCKOAhq0H9qMdz/view?usp=sharing)! Below are the instructions on how to use it.
>Doge: I swear I'm nice! 😉

```
         ▄              ▄    
        ▌▒█           ▄▀▒▌   
        ▌▒▒█        ▄▀▒▒▒▐   
       ▐▄█▒▒▀▀▀▀▄▄▄▀▒▒▒▒▒▐   
     ▄▄▀▒▒▒▒▒▒▒▒▒▒▒█▒▒▄█▒▐   
   ▄▀▒▒▒░░░▒▒▒░░░▒▒▒▀██▀▒▌   
  ▐▒▒▒▄▄▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▀▄▒▌  
  ▌░░▌█▀▒▒▒▒▒▄▀█▄▒▒▒▒▒▒▒█▒▐  
 ▐░░░▒▒▒▒▒▒▒▒▌██▀▒▒░░░▒▒▒▀▄▌ 
 ▌░▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░░░░▒▒▒▒▌ 
▌▒▒▒▄██▄▒▒▒▒▒▒▒▒░░░░░░░░▒▒▒▐ 
▐▒▒▐▄█▄█▌▒▒▒▒▒▒▒▒▒▒░▒░▒░▒▒▒▒▌
▐▒▒▐▀▐▀▒▒▒▒▒▒▒▒▒▒▒▒▒░▒░▒░▒▒▐ 
 ▌▒▒▀▄▄▄▄▄▄▒▒▒▒▒▒▒▒░▒░▒░▒▒▒▌ 
 ▐▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░▒░▒▒▄▒▒▐  
  ▀▄▒▒▒▒▒▒▒▒▒▒▒▒▒░▒░▒▄▒▒▒▒▌  
    ▀▄▒▒▒▒▒▒▒▒▒▒▄▄▄▀▒▒▒▒▄▀   
      ▀▄▄▄▄▄▄▀▀▀▒▒▒▒▒▄▄▀     
         ▀▀▀▀▀▀▀▀▀▀▀▀        
```

## Features of Doge
1. Task tracker
   - [x] Able to track different kinds of tasks (Todo, Event, Deadline)
   - [x] Mark completed tasks
   - [x] Unmark completed tasks
   - [x] Delete unwanted tasks
   - [x] List the current tasks

## Basic Commands for Doge
### 1. Adding a task:
1. **todo** `borrow book`: adds a todo task
2. **event** `read book`/`2022-02-10 20:15`: adds an event
3. **deadline** `return book`/`2022-02-12 19:10`: adds a task with a deadline


### 2. Listing all tasks:
1. **list**: generates a list of all tasks
2. **list** `<` `2022-05-20 15:20`: generates a list of tasks that falls before 20th May 2022 15:20

### 3. Deleting a task:
   1. **delete** `2`: deletes the 2nd task in the list


### 4. Marking/Unmarking a task:
   1. **mark** `1`: marks 1st task as completed
   2. **unmark** `3`: unmarks 3rd task as uncompleted


## Adding a todo task: `todo`
- Adds a todo task to the task list. 
- Format: `todo t/TASK`
```
Examples:
- todo t/read book
- todo t/eat lunch
```

## Adding an event task: `event`
- Adds an event task to the task list.
- Format: `event t/TASK d/YYYY-MM-DD o/HH:MM`
- *(Date/Time have to be valid)*
```
Examples:
- event t/read book d/2022-02-10 o/14:00
- todo t/eat lunch d/2023-05-10 o/16:00
```

## Adding a deadline task: `deadline`
- Adds an event task to the task list.
- Format: `deadline t/TASK d/YYYY-MM-DD o/HH:MM`
```
Examples:
- deadline t/read book d/2022-02-10 o/14:00
- todo t/eat lunch d/2023-05-10 o/16:00
```

## Listing certain tasks based on occurrence: `List`
- Shows a list of certain tasks in the task list.
- Format: `list l/LIMITER d/YYYY-MM-DD t/HH:MM`
- *(limiters: >, >=, =, <=, <)*
```
Examples:
- list l/< d/2022-02-10 o/14:00
- list l/>= d/2022-03-15 o/19:00
```

## Launcher class for Doge:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Implement Greet, Echo, Exit

Commit:
no message

Commit:
Revert "no message"

This reverts commit 371f15781fc9c3c115e3096a76dd45f0e49d6dd1.

fredtwt committed:
Level-1: Greet, Echo, Exit

fredtwt committed:
Delete Duke.java

fredtwt committed:
Level-2: Add, List

fredtwt committed:
A-TextUiTesting

fredtwt committed:
push updated ver to main branch

fredtwt committed:
Merge branch 'main' of https://github.com/fredtwt/ip

fredtwt committed:
updated version

fredtwt committed:
Level-5: Handle Errors

fredtwt committed:
Level-6 completed

fredtwt committed:
Level 6 completed, test cases not updated yet

fredtwt committed:
Level-6 with test cases completed

fredtwt committed:
Enum class added

fredtwt committed:
Enums applied

fredtwt committed:
Add Enum class

fredtwt committed:
Update README.md details

fredtwt committed:
Level-7 work in progress

fredtwt committed:
Add save to hard disk feature

fredtwt committed:
Add dates and times features to Doge

fredtwt committed:
Add dates and time to Deadline.java and Event.java

fredtwt committed:
Add Dates and Times to Deadline.java and Event.java

fredtwt committed:
Fix task messages

fredtwt committed:
.

fredtwt committed:
Add Dates/Times exception handlers

fredtwt committed:
Add Dates/Times exception handlers

fredtwt committed:
Add Date/Time feature, List method can now list tasks within a specific duration

fredtwt committed:
Add Date/Time feature, List method can now list tasks within a specific duration

fredtwt committed:
Add Date/Time feature, List method can now list tasks within a specific duration

fredtwt committed:
Add Date/Time feature, List method can now list tasks within a specific duration

fredtwt committed:
Fix numbering when listing task

fredtwt committed:
Fix numbering when listing task

fredtwt committed:
Implement OOP to Doge

fredtwt committed:
Separate classes into packages

fredtwt committed:
Add DeadlineException.java EventException.java ListTaskDueException.java TodoException

fredtwt committed:
Add JUnit Tests for Deadline and Parser

fredtwt committed:
Create a JAR file

fredtwt committed:
Add JavaDocs

fredtwt committed:
Add JavaDoc

fredtwt committed:
Add FindCommand.java

fredtwt committed:
Merge branch 'branch-Level-9'

fredtwt committed:
Edit README.md

fredtwt committed:
Edit JavaDocs

fredtwt committed:
Update READ.MD

fredtwt committed:
Update README.md

fredtwt committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

fredtwt committed:
Add *.xml files for checkstyle

fredtwt committed:
Include GUI for Doge

fredtwt committed:
Add DialogBox.java, MainWindow.java

fredtwt committed:
Add FXML to JavaFx

junjieteoh commented on src/main/java/doge/Parser.java:
```diff
@@ -0,0 +1,59 @@
+
+        switch (c) {
+        case "todo":
+            return new TodoCommand(new Todo(curr.substring(4).trim()));
```
I feel that it will be even better if you can abstract out the trimming process and create a variable to store it. 

ie 
`String command = curr.substring(4).trim()`

junjieteoh commented on src/main/java/doge/Storage.java:
```diff
@@ -0,0 +1,119 @@
+                    Date date;
+                    SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
+
+                    switch(curr[0].trim()) {
+                    case "T":
+                        currTask = new Todo(curr[2].trim());
+                        if (taskStatus.equals("✓")) {
+                            currTask.mark();
+                        }
+                        temp.add(currTask);
+                        break;
+                    case "D":
+                        dateTime = curr[3].trim().substring(9).trim().split(" ");
+                        date = new SimpleDateFormat("dd-MMM-yyyy").parse(dateTime[0]);
+                        currTask = new Deadline(curr[2].trim(), Doge.getDateTime(sdf.format(date) + " " + dateTime[1]));
+                        if (taskStatus.equals("✓")) {
+                            currTask.mark();
+                        }
+                        temp.add(currTask);
+                        break;
+                    case "E":
+                        dateTime = curr[3].trim().substring(3).trim().split(" ");
+                        date = new SimpleDateFormat("dd-MMM-yyyy").parse(dateTime[0]);
+                        currTask = new Event(curr[2].trim(), Doge.getDateTime(sdf.format(date) + " " + dateTime[1]));
+                        if (taskStatus.equals("✓")) {
+                            currTask.mark();
+                        }
+                        temp.add(currTask);
+                        break;
+                    }
```
The `load` method seems quite long to me. Do you think it will be possible to abstract this functionality out to another method, how about the `createLoadedTask `method

junjieteoh commented on src/main/java/doge/command/ListCommand.java:
```diff
@@ -0,0 +1,116 @@
+                    continue;
+                }
+
+                if (currDateTime != null) {
+                    switch (tempStr[0].trim()) {
+                    case "=":
+                        if (!currDateTime.isEqual(dueDateTime)) {
+                            continue;
+                        }
+                        break;
+                    case ">":
+                        if (!currDateTime.isAfter(dueDateTime)) {
+                            continue;
+                        }
+                        break;
+                    case "<":
+                        if (!currDateTime.isBefore(dueDateTime)) {
+                            continue;
+                        }
+                        break;
+                    case ">=":
+                        if (!currDateTime.isAfter(dueDateTime) || !currDateTime.isEqual(dueDateTime)) {
+                            continue;
+                        }
+                        break;
+                    case "<=":
+                        if (!currDateTime.isBefore(dueDateTime) || !currDateTime.isEqual(dueDateTime)) {
+                            continue;
+                        }
+                        break;
+                    default:
+                        throw new DogeException("Invalid limiter specified!");
+                    }
```
I will like it better if this can be abstracted out into another method that is supposed to do such duedate checking.

reignnz replied:
I agree

junjieteoh commented on src/main/java/doge/command/DeleteCommand.java:
```diff
@@ -0,0 +1,57 @@
+            throw new DogeException("Are you incapable of understanding what's an integer? Input valid task no.");
+        }
+
+        if (pos > tasks.size() || pos < 0) {
```
It seems that you are doing this position check in other subclass of the Command class too (for eg in FindCommand too), Do you think it will ne better if you abstract this out as a method in the Command class instead?

ddx-510 replied:
That seems to be a good idea to further categorize the `Command` class as commands like delete, mark, unmark requires the same integer input at `pos`. But adding to the main Command class would be troublesome as other commands do not really have the same need. So maybe creating more subclasses of Command would be a better approach?

junjieteoh reviewed (commented):
LGTM, I am very impressed to read your code. It seems that you have put in a lot of effort to make this code scalable. The documentation are written nicely and the coding convention is being strictly followed. I have tried my best to read through this multiple times and here are the suggestions I can give you. They are definitely optional but hopefully it will help you.

fredtwt committed:
Edit Doge GUI

fredtwt committed:
Add GUI to Doge

fredtwt committed:
Update README.md

ddx-510 commented on src/main/java/doge/Main.java:
```diff
@@ -0,0 +1,44 @@
+import javafx.scene.Scene;
+import javafx.scene.layout.AnchorPane;
+import javafx.stage.Stage;
+
```
Could consider including java doc for this file in the next commit :)

ddx-510 commented on src/main/java/doge/command/Command.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws DogeException;
+
+    public boolean haveTask() {
```
Just a small detail that does not matter a lot: maybe could change `haveTask()` to `hasTask()`.

ddx-510 commented on src/main/java/doge/task/Deadline.java:
```diff
@@ -0,0 +1,38 @@
+	 */
+	@Override
+	public String toString() {
+		return "D " + super.toString() + " | DEADLINE: " + this.getDateTime().format(DateTimeFormatter.ofPattern("dd-MMM" +
```
Looks like the format here is a bit off, but I do not think that is your code issue. Maybe it is due to GitHub, could you double-check it again :)

ddx-510 reviewed (commented):
LGTM, really impressed by your code quality 🥇 Just some small notes to check and fix.

reignnz commented on src/main/java/doge/DateTime.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Represents a class for parsing Date and Time
+ */
+public class DateTime {
```
I like how you encapsulated the DateTime into a separate class

reignnz replied:
I like how you encapsulated the DateTime into a separate class

reignnz commented on src/main/java/doge/Doge.java:
```diff
@@ -0,0 +1,85 @@
+    private Storage storage;
+    private Ui ui;
+    private TaskList tasks;
+    private ScrollPane scrollPane;
```
Is it possible to store the Javafx in another java file

reignnz replied:
Is it possible to store the Javafx elements in another java file?

reignnz commented on src/main/java/doge/Main.java:
```diff
@@ -0,0 +1,44 @@
+    public void start(Stage stage) {
+        try {
+            FXMLLoader fxmlLoader =
+                    new FXMLLoader(new File("./src/main/resource/view/MainWindow.fxml").toURI().toURL());
```
Could consider storing the filepath into a String

reignnz replied:
Could consider storing the filepath into a String

reignnz commented on src/main/java/doge/Parser.java:
```diff
@@ -0,0 +1,59 @@
+        String curr = input.toLowerCase().trim();
+        String c = curr.split(" ")[0].trim();
+
+        switch (c) {
```
I like how you used switch case instead of the usual if else statements

reignnz reviewed (commented):
LGTM.

fredtwt committed:
Add assertions to classes

There are several assertions to check whether the parameters of methods
are null. The files that are affected are all the subclasses of the
Command class, DateTime.java, Parser.java, Storage.java, TaskList.java,
Ui.java.

fredtwt committed:
Merge pull request #1 from fredtwt/branch-A-Assertions

Add assertions to classes

fredtwt committed:
Improve code quality in ListCommand.java

Applied Happy Path Prominent to the if else clauses, extracted the list
logic and checking date/time functionality into their own methods.

fredtwt committed:
Merge pull request #2 from fredtwt/branch-A-CodeQuality

Improve code quality in ListCommand.java

fredtwt committed:
Amend package import for ListCommand.java

The current package import in ListCommand.java is importing all files in doge.

This can lead to unnecessary file imports.

Expliciting listing only necessary imports will help us comply with coding standards.

By expliciting listing, this ensures only necessary imports are imported.

fredtwt committed:
Amend package import for ListCommand.java

fredtwt committed:
Merge branch 'master' of https://github.com/fredtwt/ip-1

fredtwt committed:
Add feature to view tasks within specified period

List currently list all tasks and not based on period.

Users would want to filter out specific tasks that are due within a specific period of time.

Implement a logic to handle for such cases when period of time is specified by user. The list command would then filter out and only list tasks that falls within that period.

The current code would then be able to handle both scenarios of listing all tasks and specific tasks.

fredtwt committed:
Improve GUI of Doge Chatbot

Tweak the GUI to match the asymmetric nature of the conversation. As the conversation is between the user and the app (not between two humans), it is asymmetric in nature. Accordingly, it makes sense not to display both sides of the conversion in the same visual format.

Changing the background colours of the user and chatbot chat bubbles and also flipping the dialog box to match the asymmetric nature of the conversation.

By changing the colours and flipping the dialog boxes, it allows the user to clearly identify and differentiate the respective chat bubbles.

fredtwt committed:
Update README.md

New features added to Doge chatbot. README.md needs to be updated to reflect the documentation of new features.

Documentation of Find feature included in README.md.

Users can refer to the documentation if they are unsure of certain features.

fredtwt committed:
Set theme jekyll-theme-slate

fredtwt committed:
Update README.md

fredtwt committed:
Amend coding standards for all files

fredtwt committed:
Merge branch 'master' of https://github.com/fredtwt/ip-1

fredtwt committed:
Edit file path for fxmlLoader

fredtwt committed:
Rename Ui.png

fredtwt committed:
Add A-CI increment
